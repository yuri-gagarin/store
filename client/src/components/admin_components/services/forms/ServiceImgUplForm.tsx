import React, { useState, useContext, useEffect } from "react";
import { Button, Form } from "semantic-ui-react";
// css imports //
import "./css/serviceImgUploadForm.css";
// actions //
import { uploadServiceImage } from "../actions/APIServiceActions";
// state //
import { Store } from "../../../../state/Store";

const ServiceImageUplForm: React.FC<{}> = (props): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, currentServiceData } = state.serviceState;
  // local file state 
  const [ file, setFile ] = useState<File>();

  // event handlers and listeners //
  const handleButtonClick = () => {

  };
  const uploadFile = () => {
    if (file) {
      const formData = new FormData();
      const { _id } = currentServiceData;
      formData.append("serviceImage", file);
      uploadServiceImage(_id, formData, state, dispatch)
        .then((_) => {
          setFile(undefined)
        })
        .catch(() => {
          // perhaps show error later //
        })
    }
  };
  const cancelFile = () => {
    setFile(undefined);
  };
  const fileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files) {
      setFile(e.target.files[0])
    }
  };
  

  return (
    <div id="serviceImgUplFormHolder">
      <div><p>Image Uploader</p></div>
      <Form id="serviceImgUploadForm">
        {
          !file ?
          <div id="serviceImgInputControlls"> 
            <Button
              id="selectServiceImgBtn"
              as="label"
              content="Choose Image"
              labelPosition="left"
              htmlFor="serviceImgFile"
              icon="file"
              onClick={handleButtonClick}
            />
            <input type="file" id="serviceImgFile" hidden onChange={fileChange} />
          </div>
          : null
        }
        
        {
          file ?  
          <div id="serviceImgUploadControlls">
            <Button 
              id="serviceImgCancelBtn"
              as="label"
              content="Cancel"
              labelPosition="left"
              icon="cancel"
              onClick={cancelFile}
            />
            <Button 
              id="serviceImgUploadBtn"
              as="label"
              content="Upload"
              icon="upload"
              onClick={uploadFile}
              loading={loading}
            />
          </div>
          : null
        }
       

       <span>{file ? file.name : "No File"}</span>
      </Form>
    </div>
  );
};

export default ServiceImageUplForm;

