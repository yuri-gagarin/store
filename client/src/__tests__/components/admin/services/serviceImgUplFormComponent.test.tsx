import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import ServiceImgUplForm from "../../../../components/admin_components/services/forms/ServiceImgUplForm";
import { Button } from "semantic-ui-react";
import MockFile from "../../../helpers/mockFile";
import moxios from "moxios";
import { createMockServices } from "../../../../test_helpers/serviceHelpers";
import { act } from 'react-dom/test-utils';

type InputProps = {
  type: string;
}
describe("Service Image Upload Form Tests", () => {

  describe("Render tests without any Image data", () => {
    let component: ShallowWrapper<React.FC>;
    beforeAll(() => {
      component = shallow<React.FC<{}>, {}>(<ServiceImgUplForm />)
     });
    it("Should properly render", () => {
      expect(component).toMatchSnapshot();
    });
    it("Should have one upload Button element", () => {
      const uplButton = component.find(Button);
      expect(uplButton.length).toEqual(1);
    });
    it("Should have an input", () => {
      const input = component.find("input");
      expect(input.length).toEqual(1);
    });
  });

  describe("Render tests with an Image file present", () => {
    let component: ShallowWrapper<React.FC>;
    let input: ShallowWrapper;
    beforeAll(() => {
      component = shallow<React.FC<{}>, {}>(<ServiceImgUplForm />);
      const file: File = MockFile.create("test", 1024, { type: "image/jpeg" });
      input = component.find("input");
      input.simulate("change", { target: { files: [ file ] } });
    });
    it("Should properly set the file in props", () => {
      expect(input.prop("type")).toBeDefined();
      expect(input.prop("type")).toEqual("file");
    });
    it("Should successfully update and render the ImgUpload button", () => {
      const imgUpoadBtn = component.find("#serviceImgUploadBtn");
      expect(imgUpoadBtn.length).toEqual(1);
    });
    it("Should successfully update and render the Cancel button", () => {
      const imgCanceldBtn = component.find("#serviceImgCancelBtn");
      expect(imgCanceldBtn.length).toEqual(1);
    });
  });
  describe("'#serviceImgUploadBtn' functionality", () => {
    let component: ShallowWrapper<React.FC>;
    let input: ShallowWrapper;

    beforeAll(() => {
      component = shallow<React.FC<{}>, {}>(<ServiceImgUplForm />);
      const file: File = MockFile.create("test", 1024, { type: "image/jpeg" });
      input = component.find("input");
      input.simulate("change", { target: { files: [ file ] } });
    });
    afterEach(() => {
      moxios.uninstall();
    });

    it("Should successfully fire the 'uploadServiceImg' action", (done) => {
      moxios.install()
      moxios.wait(() => {
        const request = moxios.requests.mostRecent();
        request.respondWith({
          status: 200,
          response: {
            responseMsg: "All ok",
            updatedService: createMockServices(1)[0]
          }
        })
      })
      const imgUpoadBtn = component.find("#serviceImgUploadBtn");
      imgUpoadBtn.simulate("click");
      setTimeout(() => {
        done()
      }, 500);
    });
    it("Should correctly render Select Image button after 'successful upload", () => {
      const selectImgBtn = component.find("#selectServiceImgBtn");
      expect(selectImgBtn.length).toEqual(1);
    });
    it("Should NOT render Image Upload button after 'successful upload", () => {
      const imgUpoadBtn = component.find("#serviceImgUploadBtn");
      expect(imgUpoadBtn.length).toEqual(0);
    })
    it("Should NOT render Cancel Upload button after 'successful upload", () => {
      const imgUpoadBtn = component.find("#serviceImgUploadBtn");
      expect(imgUpoadBtn.length).toEqual(0);
    })
  });
});