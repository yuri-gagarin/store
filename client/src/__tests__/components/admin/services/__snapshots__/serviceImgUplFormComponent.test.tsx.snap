// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Image Upload Form Tests Render tests without any Image data Should properly render 1`] = `
<div
  id="serviceImgUplFormHolder"
>
  <div>
    <p>
      Image Uploader
    </p>
  </div>
  <Form
    as="form"
    id="serviceImgUploadForm"
  >
    <div
      id="serviceImgInputControlls"
    >
      <Button
        as="label"
        content="Choose Image"
        htmlFor="serviceImgFile"
        icon="file"
        id="selectServiceImgBtn"
        labelPosition="left"
        onClick={[Function]}
      />
      <input
        hidden={true}
        id="serviceImgFile"
        onChange={[Function]}
        type="file"
      />
    </div>
    <span>
      No File
    </span>
  </Form>
</div>
`;
